File: patterns/behavioral/mediator.js
/**
 * Mediator Pattern - JavaScript
 * 
 * Defines how a set of objects interact. Promotes loose coupling by keeping
 * objects from referring to each other explicitly, and lets you vary their
 * interaction independently.
 * 
 * Use cases:
 * - Chat systems
 * - Air traffic control
 * - UI component communication
 * - Event buses
 */

// Mediator interface
class Mediator {
    notify(sender, event) {
        throw new Error('Method must be implemented');
    }
}

// Concrete mediator
class ChatMediator extends Mediator {
    constructor() {
        super();
        this.users = [];
    }
    
    addUser(user) {
        this.users.push(user);
    }
    
    notify(sender, event) {
        this.users.forEach(user => {
            if (user !== sender) {
                user.receive(event);
            }
        });
    }
}

// Colleague interface
class User {
    constructor(name, mediator) {
        this.name = name;
        this.mediator = mediator;
    }
    
    send(message) {
        console.log(`${this.name} sends: ${message}`);
        this.mediator.notify(this, { from: this.name, message });
    }
    
    receive(event) {
        console.log(`${this.name} receives from ${event.from}: ${event.message}`);
    }
}

// Example usage
console.log('=== Mediator Pattern ===\n');

const chatMediator = new ChatMediator();

const user1 = new User('Alice', chatMediator);
const user2 = new User('Bob', chatMediator);
const user3 = new User('Charlie', chatMediator);

chatMediator.addUser(user1);
chatMediator.addUser(user2);
chatMediator.addUser(user3);

user1.send('Hello everyone!');
console.log();
user2.send('Hi Alice!');
console.log();
user3.send('Hey guys!');

module.exports = { Mediator, ChatMediator, User };

File: patterns/behavioral/mediator.py
"""
Mediator Pattern - Python

Defines how a set of objects interact. Promotes loose coupling by keeping
objects from referring to each other explicitly, and lets you vary their
interaction independently.

Use cases:
- Chat systems
- Air traffic control
- UI component communication
- Event buses
"""

from abc import ABC, abstractmethod


# Mediator interface
class Mediator(ABC):
    @abstractmethod
    def notify(self, sender, event):
        pass


# Concrete mediator
class ChatMediator(Mediator):
    def __init__(self):
        self.users = []
    
    def add_user(self, user):
        self.users.append(user)
    
    def notify(self, sender, event):
        for user in self.users:
            if user != sender:
                user.receive(event)


# Colleague interface
class User:
    def __init__(self, name, mediator):
        self.name = name
        self.mediator = mediator
    
    def send(self, message):
        print(f"{self.name} sends: {message}")
        self.mediator.notify(self, {'from': self.name, 'message': message})
    
    def receive(self, event):
        print(f"{self.name} receives from {event['from']}: {event['message']}")


# Example usage
if __name__ == "__main__":
    print("=== Mediator Pattern ===\n")
    
    chat_mediator = ChatMediator()
    
    user1 = User('Alice', chat_mediator)
    user2 = User('Bob', chat_mediator)
    user3 = User('Charlie', chat_mediator)
    
    chat_mediator.add_user(user1)
    chat_mediator.add_user(user2)
    chat_mediator.add_user(user3)
    
    user1.send('Hello everyone!')
    print()
    user2.send('Hi Alice!')
    print()
    user3.send('Hey guys!')

